pallets/crowdloan-claim/src/lib.rs

```diff
@@ -790,42 +811,38 @@ impl<T: Config> Pallet<T> {
 	// contributor's claim is valid.
 	fn verify_contribution_proof(
 		relay_account: T::RelayChainAccountId,
-		proof: StorageProof,
-	) -> Result<ContributionAmountOf<T>, DispatchError> {
-		let keys = vec![relay_account.encode()];
-
-		// rebuild child info
-		let mut buf = Vec::new();
-		buf.extend_from_slice(b"crowdloan");
-		buf.extend_from_slice(
-			&Self::crowdloan_trie_index()
-				.ok_or(Error::<T>::PalletNotInitialized)?
-				.encode()[..],
-		);
-		let child_info = ChildInfo::new_default(T::Hashing::hash(&buf[..]).as_ref());
-
+		proof: Vec<T::Hash>,
+		contribution: ContributionAmountOf<T>,
+	) -> DispatchResult {
 		// We could unwrap here, as we check in the calling function if pallet is initialized (i.e. if contributions is set)
 		// but better be safe than sorry...
 		let root = Self::contributions().ok_or(Error::<T>::PalletNotInitialized)?;
 
-		read_child_proof_check::<T::Hashing, _>(root, proof, &child_info, keys).map_or(
-			Err(Error::<T>::InvalidProofOfContribution.into()),
-			|mut key_value_map| {
-				key_value_map
-					.get_mut(&relay_account.encode())
-					.map_or(&None, |map_val| map_val)
-					.as_ref()
-					.map_or(
-						Err(Error::<T>::InvalidProofOfContribution.into()),
-						|storage_val| {
-							let (value, _memo): (ContributionAmountOf<T>, Vec<u8>) =
-								Decode::decode::<&[u8]>(&mut storage_val.as_ref())
-									.unwrap_or((Zero::zero(), Vec::new()));
-
-							Ok(value)
-						},
-					)
-			},
-		)
+		// Number of proofs should practically never be > 30. Checking this
+		// blocks abuse.
+		ensure!(
+			proof.len() < T::MaxProofLength::get() as usize,
+			Error::<T>::InvalidProofOfContribution
+		);
+
+		// Concat account id : amount
+		let mut v: Vec<u8> = relay_account.encode();
+		v.extend(contribution.encode());
+
+		// Generate root hash
+		let leaf_hash = T::Hashing::hash(&v);
+		let mut root_hash = proof
+			.iter()
+			.fold(leaf_hash, |acc, hash| Self::sorted_hash_of(&acc, hash));
+
+		// Initial runs might only have trees of single leaves,
+		// in this case leaf_hash is as well root_hash
+		if proof.len() == 0 {
```

Comment:
I don't think that would be the case for us, since we will build the whole tree at once.